tests/unit/src/Internal/Service/PaymentProcessingServiceTest.php

```diff
@@ -61,4 +100,30 @@ public function test_process_payment_happy_path() {
 		$result = $this->sut->process_payment( 1 );
 		$this->assertSame( $mock_completed_state, $result );
 	}
+
+	/**
+	 * Test the basic happy path of processing a payment.
+	 */
+	public function test_process_payment_happy_path_without_mock_builder() {
+		$sut = new PaymentProcessingService( $this->mock_state_factory, $this->mock_legacy_proxy );
+
+		$mock_initial_state   = $this->createMock( InitialState::class );
+		$mock_completed_state = $this->createMock( CompletedState::class );
+
+		// Prepare a payment method for the context.
+		$_POST['payment_method']       = WC_Payment_Gateway_WCPay::GATEWAY_ID;
```

Comment:
Should we modify `$_POST` directly here? AFAICT, there are a few approaches:

- Just like this method, modify `$_POST` directly.
- Allow the request array to be provided to the `process_payment` method, which would relay it to `PaymentRequest`.
- `test_process_payment_happy_path` from this class, which mocks `create_payment_context` and `create_payment_request`. This method seems the cleanest here, but has two issues: Missing test coverage for those methods, and the fact that they need to be `protected` instead of `private`.

Which option seems best? Is there a better alternative that I'm missing?

Reply:
IMHO, modifying `$_POST` is something we should avoid. I would create a DTOish class (or pass an array as you suggested) and pass it to the process_payment method. We can even store WP request object inside the payment context, so we can use it anywhere if needed.

Reply:
I'll go with the parameter to `process_payment` approach. Keep in mind that `populate_context` might be a suitable candidate to being the one responsible for validating if all necessary data is present.

Providing the request object inside the context might hide issues until they appear later during the runtime.